servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/BeforeFinallyHttpOperator.java

```diff
@@ -142,44 +166,209 @@ public void request(final long n) {
+                                            subscriber.onComplete();
+                                            return;
+                                        }
+
+                                        final int prevState = setTerminalState();
+                                        if (prevState == TERMINATED) {
+                                            // We already cancelled and have to discard further events
+                                            return;
+                                        }
+                                        // Propagate original cancel to let Subscription observe it
+                                        final boolean propagateCancel = prevState == AWAITING_CANCEL;
+
                                         try {
-                                            if (stateUpdater.compareAndSet(ResponseCompletionSubscriber.this,
-                                                    PROCESSING_PAYLOAD, TERMINATED)) {
+                                            try {
                                                 beforeFinally.onComplete();
+                                            } catch (Throwable cause) {
+                                                subscriber.onError(cause);
+                                                return;
                                             }
-                                        } finally {
                                             subscriber.onComplete();
+                                        } finally {
+                                            cancel0(propagateCancel);
+                                        }
+                                    }
+
+                                    private int setTerminalState() {
+                                        for (;;) {
+                                            final int state = ResponseCompletionSubscriber.this.state;
+                                            assert state != IDLE;
+                                            if (state == TERMINATED) {
+                                                // We already cancelled and have to discard further events
+                                                return state;
+                                            }
+                                            if (state == PROCESSING_PAYLOAD) {
+                                                if (stateUpdater.compareAndSet(ResponseCompletionSubscriber.this,
+                                                        PROCESSING_PAYLOAD, TERMINATED)) {
+                                                    return state;
+                                                }
+                                            } else {
+                                                // re-entry, but we can terminate because this is a final event:
+                                                stateUpdater.set(ResponseCompletionSubscriber.this, TERMINATED);
```

Comment:
Perhaps slightly safer to compare and set against the value we previously read? At worst it got updated to terminated and we do an extra volatile read to discover that.